### PR TITLE
Allow Declaring Tasks with Descriptions

### PR DESCRIPTION
**Pull request by illBeRoy**
## Description
As per the [feature request and the following conversation](https://github.com/illBeRoy/taskipy/issues/34), this PR adds the ability to declare tasks explicitly:

```toml
[tool.taskipy.tasks]
test = { cmd = "python -m unittest tests/test_*.py", help = "runs all unit tests" }
```

When listing tasks, the description is displayed alongside the task:

```bash
$ poetry run task --list
test                runs all unit tests
```

Note: the old way of declaring tasks as string is still supported. Users can now choose between declaring task as strings (command only) or inline tables (command and description).

## Changes
1. Updated dependencies, including mypy that did not support python 3.9 before
2. Added new tests that cover the new functions under `TaskDescriptionTestCase`
3. Refactor: the `Task` class is now used solely for decoding tasks into standardized objects. Removed all other functionality.
4. Refactor: the `TaskRunner` class is now also in charge of: finding hooks, creating command strings, handling exit codes
5. Refactor: the `runner` value now comes from the `PyProject` class (instead of the `Task` class). This is due to the fact that the `runner` setting is global, and not nested under individual tasks (though this might be a nice feature to consider :))
6. Refactor: all logic revolving around printing the commands list was moved into a standalone `HelpFormatter` class. In addition, added coloring to the tasks list in order to better contrast between the task names and descriptions.
7. Updated the README
8. Updated `pyproject.toml` using the new description feature

### CODE REVIEW

**Comment by illBeRoy** [comment]
hey @eugenetriguba ! 
I would very much appreciate your opinion on this PR, if you have the time :)

**Comment by eugenetriguba** [comment]
@illBeRoy I'll try to give the code a full look over sometime later today or tomorrow! Here's some initial feedback I had more generally and some specific things I noticed when playing around with it in terms of the behavior.

### Discussion for v2

This is more tangential rather than something actionable for this particular feature (unless this would be a v2 release) but an initial question I had was what your thoughts were on mixing the command help text with the commands in the output. 

```
$ poetry run task --list
test                runs all unit tests
lint                 ./task lint
```

I think mixing the two is probably the right call if we're releasing this as a minor release. When this list feature was introduced, the list command was implemented to output only the commands. Therefore, this preserves that prior behavior while adding this override for help text as an option. However, I'm hesitant on whether that is the best user experience since I think the commands mixed in may start to feel like visual clutter. I may be wrong, but I figured it could be something to think.

### Inline and standard tables

After seeing it actually implemented and added to the `pyproject.toml`, I can see the argument for pushing people towards using the more verbose option or at least showing it as an option for when the lines get too long. Many of the tasks are in the 130-160 line length range, which involves quite a bit of side scrolling (for me at least). This seems quite undesirable as well.

However, there is a bit of awkwardness even if you want to convert these single line tables into standard tables. For instance, let's say the `publish_patch` task at 156 characters for its line length has become a little too long and we want to change it over to use the standard toml table. 

So instead of this:

```toml
publish_patch = { cmd = "poetry version patch && ./task publish", help = "publishes a patch version, that has only fixes but no new features (x.x.PATCH)" }
```

We convert it over to this:

```toml
[tool.taskipy.tasks.publish_patch]
cmd = "poetry version patch && ./task publish"
help = """
publishes a patch version, that has only fixes but
no new features (x.x.PATCH)
"""
```

Great! The lines aren't so long so we no longer have to do lots of side scrolling, but now we have to make another standard table for the pre and post tasks. I believe the more intuitive way to do this is would likely be to add them to the original table:

```toml
[tool.taskipy.tasks.publish_patch]
pre_cmd = "./task test"
cmd = "poetry version patch && ./task publish"
post_cmd = "echo 'some other task'"
help = """
publishes a patch version, that has only fixes but
no new features (x.x.PATCH)
"""
```

Furthermore, since no one is directly running pre and post tasks, it seems that having them in the list output is undesirable. I assume anyone running the list command would want the list of commands that are available to them to actually run.

### Text wrapping issue

It appears that for longer help texts, the current implementation isn't wrapping the text in the proposed way in https://github.com/illBeRoy/taskipy/issues/34#issuecomment-830576158. For instance, `publish_patch` has a similar following output when I run `task -l`:

```
publish_patch                   publishes a patch version, that has only fixes but no new features 
  (x.x.PATCH)
```

However, it seems the desired behavior is something more along the lines of this:

```
publish_patch                   publishes a patch version, that has only fixes but no new features 
                                (x.x.PATCH)
```

or this:

```
publish_patch                   publishes a patch version, that has only fixes but no new features 
                                  (x.x.PATCH)
```

File: taskipy/task_runner.py [hunk 1]
```diff
@@ -20,40 +23,79 @@ def list(self):
         """lists tasks to stdout"""
         longest = len(max(self.project.tasks, key=len))
         for k, v in self.project.tasks.items():
-            print(f'{k:<{longest}}  {v}')
+            print(f'{k:<{longest}}  {v.description or v.command}')
```
Comment (eugenetriguba) [inline taskipy/task_runner.py:26]: This could be a separate pull request, but I feel like it'd be more ideal to decouple this method from doing the printing and gathering of the tasks. We could just as easily return the string to print in `list()` and do a `print(runner.list())` in the `cli.py` file.

File: taskipy/task.py [hunk 2]
```diff
@@ -1,71 +1,44 @@
...
+            try:
+                return task_toml_contents['help']
+            except KeyError:
+                return ''
 
-    def __quote_arg(self, arg: str) -> str:
-        if platform.system() == 'Windows':
-            return mslex.quote(arg)
-        return shlex.quote(arg)
+        raise MalformedTaskError(self.__task_name, 'tasks must be strings, or dicts that contain { cmd, help }')
```
Comment (eugenetriguba) [inline taskipy/task.py:44]: This error message seems to convey that the tasks _must_ contain `cmd` and `help` when using a dict. However, there could just be `cmd`. I also don't know if we want to use the term "dict" for dictionary or "table" since that is what toml calls them (inline tables and standard tables).

Comment (eugenetriguba) [reply taskipy/task.py:44]: Nevermind to the table comment:

https://toml.io/en/v1.0.0#table
"Tables (also known as hash tables or dictionaries)..."

Comment (illBeRoy) [reply taskipy/task.py:44]: regarding the first part, how would you convey that `help` is optional, without making the error message too long? because, if I have to choose between not documenting "`help` being optional" and a lengthy error message, I would take the UX path here and go with the first option :)

File: taskipy/task_runner.py (hunk 1 above)
Comment (illBeRoy) [reply taskipy/task_runner.py:26]: following your comment I created a formatter that takes care of formatting the list. It does the following:
1. Indents the text
2. Handles wraps and newlines
3. Colors the command to create a clear distinction between commands and descriptions

**Comment by illBeRoy** [comment]
Hey! thanks for the insights and feedback.
Following your comments I've made some changes to the code, including the introduction of a standalone formatter for the `--list` action.

Regarding tables in the toml file: I agree with you that we should expend upon the idea of verbose tasks. Let's keep that in mind and implement this with `v2`. I'll create an RFC for `v2` in one of the following days, where we'll discuss features and improvements for v2 :)

File: taskipy/task.py (hunk 2 above)
Comment (eugenetriguba) [reply taskipy/task.py:44]: I agree with you. It's just something I figured I'd mention. It may have been better to prefix this as a general comment. I don't think it's a big deal. An option to consider could be to refer to the `{ cmd, help }` as a schema that must be followed, rather than saying contain. Your call whether you think that is better or not.

File: taskipy/exceptions.py [hunk 3]
```diff
@@ -32,6 +32,15 @@ def __init__(self, task_name: str):
     def __str__(self):
         return f'could not find task "{self.task}"'
 
+class MalformedTaskError(TaskipyError):
+    def __init__(self, task_name: str, reason: str):
+        super().__init__()
+        self.task = task_name
```
Comment (eugenetriguba) [inline taskipy/exceptions.py:38]: Do we want these to be "private"?

File: taskipy/task_runner.py [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
-from typing import List, Union
+from typing import List, Union, Optional
 
 from taskipy.pyproject import PyProject
+from taskipy.exceptions import TaskNotFoundError
 from taskipy.task import Task
+from taskipy.help import HelpFormatter
 
 
 class TaskRunner:
```
Comment (eugenetriguba) [inline taskipy/task_runner.py:17]: Same question here, whether we want `project` to be "private"?

**Comment by eugenetriguba** [comment]
Thanks for making those changes! I just had two very minor comments about whether a couple variables should be "private" or not. Regardless, I think this looks great now so thanks for doing the work for this. I like the addition of the color as well to highlight the task name. Nice work!

**Review by eugenetriguba — APPROVED** [review]

File: taskipy/task_runner.py (hunk 4 above)
Comment (illBeRoy) [reply taskipy/task_runner.py:17]: you're right! fixed that :)

File: taskipy/exceptions.py (hunk 3 above)
Comment (illBeRoy) [reply taskipy/exceptions.py:38]: I think not, as this data can be used for programmatic introspection if needed. Anyways, I was following your code when you implemented `TaskNotFoundError` ;)
